fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,541 @@
+                self.batches[chunk] = Batch(tuple(tensors), chunk)
+
+    def get_batch(self, chunk: int) -> Batch:
+        batch = self.batches[chunk]
+        assert batch is not None
+        return batch
+
+
+class PartitionHandler:
+    def __init__(
+        self,
+        module_rref: rpc.RRef,
+        device: str,
+        num_input: int,
+        num_output: int,
+        rank: int,
+        chunks: int,
+        checkpoint_stop: int,
+        loss_module_rref: Optional[rpc.RRef] = None,
+    ) -> None:
+        self.module = module_rref.local_value()
+        self.chunks = chunks
+        self.device = torch.device(device)
+        self.checkpoint_stop = checkpoint_stop
+        self.rank = rank
+        self.num_input = num_input
+        self.num_output = num_output
+        (self.in_queue,), (self.out_queue,) = create_workers([self.device])
+        self.loss_module = None if loss_module_rref is None else loss_module_rref.local_value()
+
+    def local_parameter_rrefs(self) -> List[rpc.RRef]:
+        r"""
+        Create one RRef for each parameter in the given local module, and return a
+        list of RRefs.
+        """
+        return [rpc.RRef(p) for p in self.module.parameters()]
+
+    def make_dist_record(self, users: List[Tuple[rpc.RRef, int, int]]) -> DistributedPipelineRecord:
+        return DistributedPipelineRecord(self.device, self.rank, self.chunks, self.num_input, users)
+
+    def run(self, dist_record: DistributedPipelineRecord) -> None:
+        """Runs pipeline parallelism.
+
+        It modifies the given batches in place.
+
+        """
+
+        m = len(dist_record.batches)
+
+        self.stream = current_stream(self.device)
+
+        for i in range(m):
+            with record_function("feed"):
+                dist_record.wait_for(i)
+            self.fence(dist_record, i)
+            self.compute(dist_record, i)
+            self.forward_results(i, dist_record)
+
+    def fence(self, dist_record: DistributedPipelineRecord, chunk: int) -> None:
+        """Copies micro-batches after computation for the previous
+        micro-batches.
+        """
+        # Ensure that batches[chunk-1] is executed after batches[chunk] in
+        # backpropagation by an explicit dependency.
+        if chunk != 0 and dist_record.users and dist_record.rank > 1:
+            t = []
+            batch = dist_record.batches[chunk]
+            assert batch is not None
+            for b, remote_ph_list in zip(batch.tensors, dist_record.forwarded_phony[chunk - 1]):
+                r = b
+                for remote_ph in remote_ph_list:
+                    ph = remote_ph.to_here()
+                    r = join(r, ph)
+                t.append(r)
+            dist_record.batches[chunk] = Batch(tuple(t), chunk)
+
+    def compute(self, dist_record: DistributedPipelineRecord, chunk: int) -> None:
+        """Runs tasks with synchronization to copy streams."""
```

Comment:
This comment looks wrong. I don't see any use of copy streams.

Reply:
Updated the comment. It is now synchronization to tensorpipe steams (with wait_event calls).